src/bgw/job.c

```diff
@@ -723,11 +723,23 @@ get_job_lock_for_delete(int32 job_id)
 			proc = BackendIdGetProc(vxid->backendId);
 			if (proc != NULL && proc->isBackgroundWorker)
 			{
-				elog(NOTICE,
-					 "cancelling the background worker for job %d (pid %d)",
-					 job_id,
-					 proc->pid);
-				DirectFunctionCall1(pg_cancel_backend, Int32GetDatum(proc->pid));
+				/* Simply assuming that this pid corresponds to the background worker
+				 * running the job is not sufficient. The scheduler could also be the
+				 * one holding the lock, when transitioning the state of the job back
+				 * to scheduled state. So we must check we don't kill the scheduler.
+				 * See https://github.com/timescale/timescaledb/issues/5224
+				 */
+				const char *worker_name = GetBackgroundWorkerTypeByPid(proc->pid);
+				const char *bgw_scheduler_name = "TimescaleDB Background Worker Scheduler";
+				/* do not kill the scheduler */
+				if (strcmp(worker_name, bgw_scheduler_name) != 0)
```

Comment:
I saw this during test runs - I see that its still in draft but wanted to note that:

this doesn't look 100% right to me

I believe this method wanted to provide a postcondition of `the job is not running anymore`  ; however with this fix  that will not be true anymore in all cases - as it may just mask it

is this the final fix?

Reply:
I understand that it might be in transistioning state for this job - but from that I don't think its implied that its being "put back" into the `SCHEDULED` state - could it be transitioning into something else?

on the other hand I agree that...this will remove the most serious issue - meaning the kill of the schq thread